openml/datasets/functions.py

```diff
@@ -419,6 +443,15 @@ def get_dataset(
             "`dataset_id` must be one of `str` or `int`, not {}.".format(type(dataset_id))
         )
 
+    # Note: we could also (quite heavily) re-implement the below to only download the
+    # data and do not cache the data at all. This would always be thread/multiprocessing
+    # safe. However, this would likely drastically increase the strain on the server.
```

Comment:
And make any work with OpenML really slow. Please see my comment above on how to make this class thread-safe.

Reply:
+1, we definitely want to cache data. In the vast amount of use cases it is more important than making this function thread safe.